### PR TITLE
refactor(build): include lpeg as a library

### PR DESCRIPTION
**Pull request by bfredl**
This reduces the _build time_ required luarocks packages to zero. (luarocks is now only needed for testing).

alternative: vendor lpeg as `src/lpeg` just like `src/mpack`. This avoids most extra cmake code, but vendors the lpeg c code (~5000 lines). Not sure which one will make distros recoil in horror the most.

### CODE REVIEW

**Comment by clason** [comment]
Do we want to make changes to `lpeg` for _any_ reason? That is usually the deciding factor on whether to vendor something.

**Comment by clason** [comment]
On the other hand, I'm always wary of using external download sites that are not obviously part of a CDN. Maybe we could host a mirror on Github as a first step to guard against download failures?

File: runtime/doc/lua.txt [hunk 1]
```diff
@@ -648,6 +648,13 @@ regex:match_line({bufnr}, {line_idx} [, {start}, {end}])  *regex:match_line()*
     |regex:match_str()|. If {start} is used, then the returned byte indices
     will be relative {start}.
 
+------------------------------------------------------------------------------
+VIM.LPEG                                                            *lua-lpeg*
+
+                                                             *vim.lpeg* *vim.re*
+The Lpeg library for parsing expression grammars is being included as
+`vim.lpeg`. In additional its regex-like interface is available as `vim.re`.
```
Comment (clason) [inline runtime/doc/lua.txt:656]:
```suggestion
`vim.lpeg`. In addition, its regex-like interface is available as `vim.re`.
```

File: runtime/doc/lua.txt [hunk 2]
```diff
@@ -648,6 +648,13 @@ regex:match_line({bufnr}, {line_idx} [, {start}, {end}])  *regex:match_line()*
     |regex:match_str()|. If {start} is used, then the returned byte indices
     will be relative {start}.
 
+------------------------------------------------------------------------------
+VIM.LPEG                                                            *lua-lpeg*
+
+                                                             *vim.lpeg* *vim.re*
+The Lpeg library for parsing expression grammars is being included as
+`vim.lpeg`. In addition, its regex-like interface is available as `vim.re`.
```
Comment (clason) [inline runtime/doc/lua.txt:656]:
```suggestion
`vim.lpeg`. In addition, its regex-like interface is available as `vim.re`, see https://www.inf.puc-rio.br/~roberto/lpeg/re.html for documentation.
```

Or do we want to vendor that as well?

File: cmake/FindLpeg.cmake [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+function(ubuntu)
+  set(CMAKE_FIND_LIBRARY_PREFIXES "")
+  find_library(LPEG_LIBRARY NAMES lpeg PATH_SUFFIXES lua/5.1)
+endfunction()
+ubuntu()
+
+message(STATUS "KRARK")
+message(STATUS "${LPEG_LIBRARY}")
+
+find_package_handle_standard_args(Lpeg DEFAULT_MSG LPEG_LIBRARY )
```
Comment (dundargoc) [inline cmake/FindLpeg.cmake:13]: formatting

File: cmake.deps/cmake/LpegCMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.10)
+project (lpeg C)
+
+include(GNUInstallDirs)
+
+file(GLOB LPEG_SOURCES ${CMAKE_SOURCE_DIR}/*.c)
+add_library(lpeg STATIC ${LPEG_SOURCES})
```
Comment (dundargoc) [inline cmake.deps/cmake/LpegCMakeLists.txt:7]: STATIC isn't needed, we're installing all deps as STATIC by default.

File: runtime/doc/lua.txt (hunk 2 above)
Comment (bfredl) [reply runtime/doc/lua.txt:656]: maybe but it could be a followup. adding the links for now.

**Review by dundargoc — APPROVED** [review]
The cmake part looks fine apart from some nitpicks.

File: runtime/doc/lua.txt (hunk 2 above)
Comment (clason) [reply runtime/doc/lua.txt]: (I looked at the docs; they're not huge, so pulling them in as `lua.txt` is perfectly feasible, and possibly useful -- depending on how useful `vim.lpeg` and `vim.re` is for users, of course. But as you say, possible followup, and good-first-issue.)

**Comment by clason** [comment]
Need to add `runtime/lua/vim/re.lua` to `.styluaignore` and `.luacheckrc`.

**Review by clason — APPROVED** [review]

**Pull request merged**